cmd/jaeger/internal/extension/jaegerstorage/extension.go

```diff
@@ -60,35 +62,62 @@ func newStorageExt(config *Config, otel component.TelemetrySettings) *storageExt
 	}
 }
 
-func (s *storageExt) Start(ctx context.Context, host component.Host) error {
-	for name, mem := range s.config.Memory {
-		if _, ok := s.factories[name]; ok {
-			return fmt.Errorf("duplicate memory storage name %s", name)
-		}
-		s.factories[name] = memory.NewFactoryWithConfig(
-			mem,
-			metrics.NullFactory,
-			s.logger.With(zap.String("storage_name", name)),
-		)
-	}
+type starter[Config any, Factory storage.Factory] struct {
+	ext         *storageExt
+	storageKind string
+	cfg         map[string]Config
+	builder     func(Config, metrics.Factory, *zap.Logger) (Factory, error)
+}
 
-	for name, b := range s.config.Badger {
-		if _, ok := s.factories[name]; ok {
-			return fmt.Errorf("duplicate badger storage name %s", name)
+func (s *starter[Config, Factory]) build(ctx context.Context, host component.Host) error {
+	for name, cfg := range s.cfg {
+		if _, ok := s.ext.factories[name]; ok {
+			return fmt.Errorf("duplicate %s storage name %s", s.storageKind, name)
 		}
-		var err error
-		factory, err := badger.NewFactoryWithConfig(
-			b,
+		factory, err := s.builder(
+			cfg,
 			metrics.NullFactory,
-			s.logger.With(zap.String("storage_name", name)),
+			s.ext.logger.With(zap.String("storage_name", name)),
 		)
 		if err != nil {
-			return fmt.Errorf("failed to initialize badger storage: %w", err)
+			return fmt.Errorf("failed to initialize %s storage %s: %w", s.storageKind, name, err)
 		}
-		s.factories[name] = factory
+		s.ext.factories[name] = factory
 	}
+	return nil
+}
 
-	// TODO add support for other backends
+func (s *storageExt) Start(ctx context.Context, host component.Host) error {
+	memStarter := &starter[memoryCfg.Configuration, *memory.Factory]{
+		ext:         s,
+		storageKind: "memory",
+		cfg:         s.config.Memory,
+		// memory factory does not return an error, so need to wrap it
```

Comment:
Why not change `memory.NewFactoryWithConfig` to return the error? It doesn't look like it's used anywhere else, so it shouldn't be a large breaking change.

Reply:
Because I hate functions that are declared with errors but always return nil. The caller either has to ignore the error, or it has no easy way of unit testing the error handing path (which will never be used). Lose-lose. I think if the function can never return an error it should never declare it as a possibility (unless you're dealing with an interface where there's no choice).